packages/splitter/src/useSplitter.ts

```diff
@@ -0,0 +1,274 @@
+
+export enum SplitterOrientation {
+  HORIZONTAL = 'horizontal',
+  VERTICAL = 'vertical'
+}
+
+export enum SplitterType {
+  FIXED = 'fixed',
+  VARIABLE = 'variable'
+}
+export interface IUseSplitterProps extends Omit<HTMLProps<any>, 'onChange' | 'controls'> {
+  /** A label for the separator */
+  label?: string;
+  /** An id identifying the element labeling the separator */
+  labelledBy?: string;
+  /** An id of the element the separator controls */
+  controls: string;
+  /** A document environment to attach events to */
+  environment?: Pick<Document, 'addEventListener' | 'removeEventListener'>;
+  /** A default value for starting uncontrolled separator location */
+  defaultValueNow?: number;
+  /** Determines whether a separator behaves in fixed or variable mode */
+  type: SplitterType;
+  /** A minimum number for separator location */
+  min: number;
+  /** A maximum number for separator location */
+  max: number;
+  /** Determines whether the separator is vertical or horizontal orientation */
+  orientation: SplitterOrientation;
+  /** Determines current position of a controlled separator */
+  valueNow?: number;
+  /** A callback that receives updated valueNow when a controlled separator moves */
+  onChange?: (valueNow: number) => void;
+}
+
+export interface IUseSplitterReturnValue {
+  getSeparatorProps: <T>(options?: T & HTMLProps<any>) => any;
+}
+
+export interface IUsePrecachedStateProps<T> {
+  /* A default value for uncontrolled mode starting value */
+  defaultValue: T;
+  /* A current value for controlled state mode */
+  value?: T;
+  /* A callback that recieves updated value when controlled state changes */
+  onChange?: (value: T) => void;
+}
+
+// Problem: Generic DOM events fire independently from React state lifecycle.
+// Sometimes this will cause React state changes driven by global events to lag behind the current DOM value
+// usePrecachedState will provide "safe" state handling using a pre-cached Ref when listening to global DOM events to keep up with latest changes.
+// Eventually React State will catch up and the virtual DOM will be consistent with the DOM when using this state hook.
+// TODO Promote to Container Utilities?
+export const usePrecachedState = <T>({
+  defaultValue,
+  value,
+  onChange
+}: IUsePrecachedStateProps<T>): [T, (value: T) => void, React.RefObject<T>] => {
+  const isControlled = value !== undefined && value !== null && typeof onChange === 'function';
+  const defaultStateValue = isControlled ? value : defaultValue;
+  const [state, setState] = useState<T>(defaultStateValue);
+  const preCacheRef = useRef<T>(defaultStateValue);
+
+  const makeCachedSetter = useCallback(
+    (setterMethod: (value: T) => void) => (nextValue: T) => {
+      setterMethod(nextValue);
+      // use a ref to cache the immediate value transitions as React state lags slightly behind the global events
+      preCacheRef.current = nextValue;
```

Comment:
React state lags behind the global event loop. Therefore use a Ref to pre-cache the eventual React State update. Even with this optimization, we need to brainstorm a way to further optimize the FPS for this tool as I noticed a slowdown when adding more CSS.